### ISSUE TITLE
Add utilities for analyzing ontology terms and identifiers in a biopax file

### ISSUE THREAD

**Issue by cmungall** [indralab/pybiopax#34]
I would like to take a biopax file and get a report of

1. which ID prefixes are used, how often, and are they well-formed
2. of the IDs that reference ontology terms, are these terms valid, in the expected branch of the ontology, etc

1 is easily achieved with bioregistry, for 2 there are various ways, obviously I'd use [OAK](https://github.com/INCATools/ontology-access-kit/)

If I was to write this, would this be a PR on this repo or a new repo/pypi package that imports pybiopax? If the former, would we add the dependencies in the setup.py or do this using 'extras' as per how OAK uses gilda?

The answer may be a hybrid - e.g. adding bioregistry as a dependency is lightweight, but OAK is a more heavyweight dependency so this may not be added to setup.py

**Comment by bgyori** [comment]
Hi @cmungall, I can implement a feature here in pybiopax, which, given a PyBioPAX model, collects and returns a list of prefix/identifier pairs used in the model. The rest of the code, i.e., checking whether prefixes and identifiers are valid are best implemented outside the core pybiopax package. We have a notebooks folder at https://github.com/indralab/pybiopax/tree/master/notebooks where we could put custom analysis like this, or it could live in another package.

**Comment by bgyori** [comment]
A preliminary example:

```python
import pybiopax
model = pybiopax.model_from_owl_gz('pybiopax/tests/biopax_test.owl.gz')
refs = model.get_objects_by_type(pybiopax.biopax.Xref)
Counter([r.db for r in refs]).most_common()
```

which reveals the ... how do I put this ... impressive diversity of prefixes used:

```
[('GENPEPT', 6025),
 ('UNIPROT KNOWLEDGEBASE', 4857),
 ('NUCLEOTIDE GENBANK IDENTIFIER', 4435),
 ('PROTEIN DATA BANK', 3328),
 ('GENE ONTOLOGY', 2382),
 ('Ensembl', 1805),
 ('PUBMED', 1649),
 ('REACTOME', 1508),
 ('UniProt Knowledgebase', 1240),
 ('ENSEMBL', 1028),
 ('KEGG REACTION', 881),
 ('RefSeq', 776),
 ('UNIPROT', 772),
 ('REFSEQ', 749),
 ('HGNC', 704),
 ('BIOGRID', 562),
 ('NCBI GENE', 501),
 ('OMIM', 482),
 ('INTERPRO', 439),
 ('GENBANK INDENTIFIER', 419),
 ('UniProt Isoform', 360),
 ('HGNC Symbol', 287),
 ('NCBI Gene', 277),
 ('HPRD', 255),
 ('VEGA', 232),
 ('HGNC SYMBOL', 220),
 (None, 215),
 ('PHARMGKB GENE', 200),
 ('PANTHER FAMILY', 152),
 ('PubMed', 142),
 ('ChEBI', 120),
 ('UNIPROT ACCESSION', 116),
 ('INTACT', 99),
 ('GENBANK GENE DATABASE', 97),
 ('HUGO GENE NOMENCLATURE COMMITTEE (HGNC)', 96),
 ('GENATLAS', 87),
 ('GENECARDS', 81),
 ('GENBANK PROTEIN DATABASE', 79),
 ('PANTHER PATHWAY COMPONENT', 65),
 ('CHEBI', 61),
 ('REACTOME DATABASE ID RELEASE 50', 55),
 ('PROTEIN MODIFICATION ONTOLOGY', 53),
 ('PUBCHEM-COMPOUND', 38),
 ('CHEMICAL NAME', 31),
 ('InChIKey', 31),
 ('CAS', 28),
 ('KEGG COMPOUND', 24),
 ('HETEROGEN', 20),
 ('Protein Modification Ontology', 18),
 ('MOLECULAR INTERACTIONS ONTOLOGY', 16),
 ('CHEMBL', 15),
 ('PUBCHEM-SUBSTANCE', 13),
 ('KEGG GENES', 10),
 ('AMAZE', 9),
 ('UNIPROT ISOFORM', 8),
 ('3DMET', 8),
 ('CHEMICAL COMPONENT DICTIONARY', 8),
 ('DrugBank', 8),
 ('KEGG DRUG', 7),
 ('EVIDENCE CODE ONTOLOGY', 7),
 ('RAT GENOME DATABASE', 6),
 ('ENZYME NOMENCLATURE', 5),
 ('LIPIDBANK', 4),
 ('KEGG GLYCAN', 4),
 ('Wikipedia', 4),
 ('CHEMSPIDER', 3),
 ('IUPHAR', 2),
 ('TAXONOMY', 2),
 ('GUIDE TO PHARMACOLOGY', 2),
 ('BINDINGDB', 2),
 ('MI', 2),
 ('PHARMGKB DRUG', 1),
 ('DRUGBANK', 1),
 ('ENA', 1),
 ('MOLECULAR MODELING DATABASE', 1),
 ('Molecular Interactions Ontology', 1)]

```

**Comment by cmungall** [comment]
That is an impressive set... though I am not sure all are "prefixes" in the bioregistry sense, as often (but not always) the whole CURIE is represented in biopax ID.

But anyway, thanks for the examples very useful, it seems like another package might be best!

**Comment by bgyori** [comment]
Well, each of these appear as the `db` attribute of a reference object (any of PublicationXref, UnificationXref or RelationshipXref) which I think ought to be equivalent to prefixes in the Bioregistry sense. Separately, for some namespaces like CHEBI, the namespace is embedded in the identifier (i.e., the `id` attribute of the reference object) as is the convention in identifiers.org for some namespaces.

In #35, I looked at some examples coming from the latest version of PathwayCommons, and the prefixes look a bit better there. See this example notebook I merged as part of that PR for a report of prefix/ID validity: https://github.com/indralab/pybiopax/blob/master/notebooks/Reference%20validity.ipynb
